src/material-experimental/mdc-form-field/form-field.ts

```diff
@@ -0,0 +1,462 @@
+
+  /** The color palette for the form-field. */
+  @Input() color: ThemePalette = 'primary';
+
+  /** Whether the label should always float or float as the user types. */
+  @Input()
+  get floatLabel(): FloatLabelType {
+    return this._floatLabel || (this._labelOptions && this._labelOptions.float)
+        || DEFAULT_FLOAT_LABEL;
+  }
+  set floatLabel(value: FloatLabelType) {
+    if (value !== this._floatLabel) {
+      this._floatLabel = value;
+      // For backwards compatibility. Custom form-field controls or directives might set
+      // the "floatLabel" input and expect the form-field view to be updated automatically.
+      // e.g. autocomplete trigger. Ideally we'd get rid of this and the consumers would just
+      // emit the "stateChanges" observable. TODO(devversion): consider removing.
+      this._changeDetectorRef.markForCheck();
+    }
+  }
+  private _floatLabel: FloatLabelType;
+
+  /** The form-field appearance style. */
+  @Input()
+  get appearance(): MatFormFieldAppearance { return this._appearance; }
+  set appearance(value: MatFormFieldAppearance) {
+    this._appearance = value || (this._defaults && this._defaults.appearance) || DEFAULT_APPEARANCE;
+  }
+  private _appearance: MatFormFieldAppearance = DEFAULT_APPEARANCE;
+
+  /** Text for the form field hint. */
+  @Input()
+  get hintLabel(): string { return this._hintLabel; }
+  set hintLabel(value: string) {
+    this._hintLabel = value;
+    this._processHints();
+  }
+  private _hintLabel = '';
+
+  // Unique id for the hint label.
+  _hintLabelId: string = `mat-hint-${nextUniqueId++}`;
+
+  // Unique id for the internal form field label.
+  _labelId = `mat-form-field-label-${nextUniqueId++}`;
+
+  /** Whether the Angular animations are enabled. */
+  _animationsEnabled: boolean;
+
+  /** State of the mat-hint and mat-error animations. */
+  _subscriptAnimationState: string = '';
+
+  /** Gets the current form field control */
+  get _control(): MatFormFieldControl<any> {
+    return this._explicitFormFieldControl || this._formFieldControl;
+  }
+  set _control(value) { this._explicitFormFieldControl = value; }
+
+  private _destroyed = new Subject<void>();
+  private _isFocused: boolean|null = null;
+  private _explicitFormFieldControl: MatFormFieldControl<any>;
+  private _foundation: MDCTextFieldFoundation;
+  private _adapter: MDCTextFieldAdapter = {
+    addClass: className => this._textField.nativeElement.classList.add(className),
+    removeClass: className => this._textField.nativeElement.classList.remove(className),
+    hasClass: className => this._textField.nativeElement.classList.contains(className),
+
+    hasLabel: () => this._hasFloatingLabel(),
+    isFocused: () => this._control.focused,
+    hasOutline: () => this._hasOutline(),
+
+    // Handled through template input binding. This gives us more flexibility. e.g. by
+    // respecting "floatLabel" @Input and supporting the "always floating" mode.
+    floatLabel: () => {},
```

Comment:
Updated the comment. Should be more clear now. basically this method only fires in certain cases.. but this is not sufficient since we want to float/dock the label whenever the `@Input() floatLabel` changes, or if the `MatFormFieldControl#shouldLabelFloat` property changes.